### DISCUSSION TITLE
Can the tgslurmbot be generalized to send messages to other users/groups?

### DISCUSSION THREAD

**Discussion by hariseldon99** [CLIP-HPC/goslmailer#33] [Ideas]
Hi,

I'm running goslmailer with tgslurmbot configged. As of now, the bot seems to only be able to communicate to its own chatid. I have a telegram group for my cluster where downtime/uptime announcements are made by another bot (basically an ugly python hack that [I cooked up here](https://github.com/hariseldon99/buparamshavak/blob/main/shavak_root/usr/local/bin/tgbot.py)). I'd like to have tgslurmbot do it, but I don't know the go language well enough to figure out how.

**Comment by hariseldon99** [comment]
Hi,

Sorry for the hasty post. I actually figured out how to do it with http requests:


```python
import requests
TOKEN = "YOUR TELEGRAM BOT TOKEN"
chat_id = "YOUR CHAT ID"
message = "hello from your telegram bot"
url = f"[https://api.telegram.org/bot{TOKEN}/sendMessage?chat_id={chat_id}&text={message](https://api.telegram.org/bot%7BTOKEN%7D/sendMessage?chat_id={chat_id}&text={message)}"
print(requests.get(url).json()) # this sends the message
```

**Comment by pja237** [comment]
Hey,
goslmailer telegram connector should be capable of sending messages to group chats same as to individual chats.
In your group, search and add the bot as the member and he should reply to `/start` and send messages about the job to the whole group.

Try it out and let me know if that is what you needed? (we tested this functionality long time ago and it worked., i hope it still works, unless telegram api changed in the meantime, then we'll update the code)

**Reply by hariseldon99** [reply]
Hi.

If I give the slurm jobscript the chatid of the group, there is an error in the logs


```bash
$ grep mail cupy_test_mps.sbatch 
#SBATCH --mail-type=ALL
#SBATCH --mail-user="telegram:-1001215703471"
```

log:

```bash
goslmailer:2023/02/28 18:01:46.864562 telegram.go:54: ................... sendToTelegram START ........................................
goslmailer:2023/02/28 18:01:46.864567 connector_data.go:19: telegram.dumpConnector: name: "telegram bot"
goslmailer:2023/02/28 18:01:46.864577 connector_data.go:20: telegram.dumpConnector: url: ""
goslmailer:2023/02/28 18:01:46.864582 connector_data.go:21: telegram.dumpConnector: token: "6246165030:AAFt92FHQtakoYUrFyV_1fDoIwciWcmR-rk"
goslmailer:2023/02/28 18:01:46.864585 connector_data.go:22: telegram.dumpConnector: renderToFile: "no"
goslmailer:2023/02/28 18:01:46.864590 connector_data.go:23: telegram.dumpConnector: spoolDir: "/var/spool/telegramgobs"
goslmailer:2023/02/28 18:01:46.864594 connector_data.go:24: telegram.dumpConnector: messageTemplate: "/etc/slurm/telegramTemplate.md"
goslmailer:2023/02/28 18:01:46.864597 connector_data.go:25: telegram.dumpConnector: useLookup: "no"
goslmailer:2023/02/28 18:01:46.864601 connector_data.go:26: telegram.dumpConnector: format: "MarkdownV2"
goslmailer:2023/02/28 18:01:46.864605 connector_data.go:27: ................................................................................
goslmailer:2023/02/28 18:01:46.864609 telegram.go:65: Looked up with "no" -1001215703471 -> -1001215703471
goslmailer:2023/02/28 18:01:46.864614 telegram.go:67: Sending to targetUserID: -1001215703471
goslmailer:2023/02/28 18:01:47.877953 telegram.go:126: chatbyusername failed telegram: chat not found (400)
goslmailer:2023/02/28 18:01:47.878162 goslmailer.go:98: ERROR in telegram.SendMessage(): "telegram: chat not found (400)"
goslmailer:2023/02/28 18:01:47.878427 goslmailer.go:103: ========================== END OF RUN ==========================================
```

If I replace it with the chatid of the bot itself, it responds to me privately, but does not post to the group.

**Comment by pja237** [comment]
I'm sorry, i can not reproduce this, i've just now created this minimal config and the bot sends the message to the group chat as expected:

goslmailer.conf:
```
[root@ctl-0 slurm]# cat goslmailer.conf 
{
  "logfile": "/tmp/goslmailer.log",
  "debugconfig": true,
  "binpaths": {
    "sacct":"/bin/sacct",
    "sstat":"/bin/sstat"
  },
  "defaultconnector": "telegram",
  "connectors": {
    "telegram": {
      "name": "pjtestbot",
      "url": "",
      "token": "REDACTED",
      "renderToFile": "no",
      "spoolDir": "/tmp/telegramgobs",
      "messageTemplate": "/etc/slurm/telegramTemplate.html",
      "useLookup": "no",
      "format": "HTML"
    }
  },
  "qosmap": {
    "elevated": 3600,
    "normal": 28800
  }
}
```

goslmailer.log:
```
[root@ctl-0 slurm]# grep 'telegram.go' /tmp/goslmailer.log 
goslmailer:2023/02/28 09:52:52.385781 telegram.go:54: ................... sendToTelegram START ........................................
goslmailer:2023/02/28 09:52:52.385823 telegram.go:65: Looked up with "no" -949039698 -> -949039698
goslmailer:2023/02/28 09:52:52.385826 telegram.go:67: Sending to targetUserID: -949039698
goslmailer:2023/02/28 09:52:52.605917 telegram.go:138: bot.Send() successful, messageID: 45
goslmailer:2023/02/28 09:52:52.605981 telegram.go:153: ................... sendToTelegram END ..........................................
goslmailer:2023/02/28 09:53:50.323857 telegram.go:54: ................... sendToTelegram START ........................................
goslmailer:2023/02/28 09:53:50.323886 telegram.go:65: Looked up with "no" -949039698 -> -949039698
goslmailer:2023/02/28 09:53:50.323889 telegram.go:67: Sending to targetUserID: -949039698
goslmailer:2023/02/28 09:53:50.580867 telegram.go:138: bot.Send() successful, messageID: 46
goslmailer:2023/02/28 09:53:50.580914 telegram.go:153: ................... sendToTelegram END ..........................................
```

Did you __add__ the bot to the group? He __must__ be a part of the group to be able to send there. (afair)

EDIT: also, i've tested out the telegramTemplate.html from the release zip and that also seems to work as expected. Will close the #31 then.

**Reply by hariseldon99** [reply]
My cluster is down for maintainance. I'll post with updates in a few hours when it is back up.

In the meantime, AFAIK I did add the bot to the group, but I'll check the ID again to be sure that it's the right one.

**Update:** I used curl to send a message to the telegram group with the api key of the bot. It worked!

```bash
curl -X POST "https://api.telegram.org/bot<API-KEY-WUZ-HERE>/sendMessage" -d "chat_id=-1001215703472&text=Herp! Derp!"
``` 

However, I was unable to get slurm to do it. I'll update on this when the machine is back up.

**Reply by pja237** [reply]
ok, when the cluster is up, please check that:
1. make sure the correct bot/token is __added__ to the group/config
2. the chatId you're using in slurm --mail-user=telegram:cid , came from that same bots reply to /start command in that group

**Reply by hariseldon99** [reply]
Hmm. It seems you were right. It's working now. I had to remove quotes from the command flag.

~~#SBATCH --mail-user="telegram:-1001215703472"~~

```bash
#SBATCH --mail-user=telegram:-1001215703472
```

Thanks.
